keps/sig-storage/1710-selinux-relabeling/README.md

```diff
@@ -51,223 +48,183 @@
 
-Optionally (chosen by user), do not recursively relabel content of the volumes.
+## Motivation
 
-### Non-Goals
+* File relabeling in CRI can be slow for big volumes.
+* Avoiding out-of-space issues when relabeling almost full volumes. When a volume is almost full, CRI can fail to relabel volumes on it, since SELinux labels may need some little space on the volume. 
+* Allowing access to read-only volumes. A read-only volume can still be mounted with `-o context=XYZ` and provide files with the right labels to a Pod.
+* Mounting volumes with the right SELinux context is a bit safer. Consider [CVE-2021-25741](https://access.redhat.com/security/cve/cve-2021-25741) - here a user can cause Kubernetes to provide host's filesystem to an innocent pod. Without this KEP, CRI will actually relabel the host's files so the Pod can access them. With this KEP, the attacker could still fool Kubernetes to mount host filesystem to the Pod, but the pod would not be able to access it, because (unprivileged) pods are denied accessing files on the host due to SELinux policy.
 
-Change container runtimes / CRI.
 
-## Proposal
+### Goals
 
-Offer option in `Pod.Spec.PodSecurityContext to` *mount* volumes with the right labels instead of recursive relabeling:
+* If possible, mount volumes with the correct SELinux context using `-o context=XYZ` mount option and avoid recursive change of all files on the volume.
 
-```go
-type SELinuxRelabelPolicy string
+### Non-Goals
 
-const (
-    OnVolumeMount SELinuxRelabelPolicy = "OnVolumeMount"
-    AlwaysRelabel SELinuxRelabelPolicy = "Always"
-)
+* Change container runtimes / CRI.
 
-type PodSecurityContext struct {
-    // seLinuxRelabelPolicy ← new field
-    // Defines behavior of changing SELinux labels of the volume before being exposed inside Pod.
-    // Valid values are "OnVolumeMount" and "Always". If not specified, "Always" is used.
-    // "Always" policy recursively changes SELinux labels on all files on all volumes used by the Pod.
-    // "OnVolumeMount" tries to mount volumes used by the Pod with the right context and skip recursive ownership
-    // change. Kubernetes may fall back to policy "Always" if a storage backed does not support this policy.
-    // This field is ignored for Pod's volumes that do not support SELinux.
-    // + optional
-    SELinuxRelabelPolicy *SELinuxRelabelPolicy
+## Proposal
 
+* When kubelet *knows* SELinux context of a pod (i.e. at least `Pod.Spec.SecurityContext.SELinuxOptions.Level` is set) AND kubelet *knows* that a volume supports mounting with `-o context`:
+  * kubelet passes `-o context=XYZ` to `MountDevice()` and `SetUp()` calls of the volume.
+    * A volume plugin / CSI driver will get these as regular mount options and use them to mount the volume.
+  * kubelet does not pass any special SELinux option to the container runtime (explicitly, no `:Z`).
+    Files on the volume already have the right SELinux context, no recursive relabeling happens there.
```

Comment:
I think this split responsibility about who should relabel what can be a concern. but from a container runtime perspective this is fine.